features/steps/thens/relations.py

```diff
@@ -40,6 +39,51 @@ def step_impl(context, entity, condition, directness, other_entity):
 
     misc.handle_errors(context, errors)
 
+@then('Each {entity} must be {relationship} {preposition} other {other_entity} as per {table}')
+def step_impl(context, entity, relationship, preposition, other_entity, table):
+
+    stmt_to_op = {'aggregated': 'Decomposes'}
+    assert relationship in stmt_to_op
+
+    tbl_path = system.get_abs_path(f"resources/{table}")
+    tbl = system.get_csv(tbl_path, return_type='dict')
+
+    ent_tbl_header, relationship_tbl_header = list(tbl[0].keys())
+
+    aggregated_table = {}
+
+    for d in tbl:
+        applicable_entity = d[ent_tbl_header]
+        tbl_relationship_object = d[relationship_tbl_header]
+        if applicable_entity in aggregated_table:
+            aggregated_table[applicable_entity].append(tbl_relationship_object)
+        else:
+            aggregated_table[applicable_entity] = [tbl_relationship_object]
+    errors = []
+
+    if context.instances and getattr(context, 'applicable', True):
+        for ent in context.model.by_type(entity):
+            for applicable_entity in aggregated_table.keys(): # check which applicable entity the currently processed entity is (inheritance), e.g IfcRailway -> IfcFacility
+                if ent.is_a(applicable_entity):
+                    break # TODO - > is the hierarchy important?
```

Comment:
What the code does here is essentially running through the dict similar to the one below (simplified):

```
{IfcFacilityPartCommon : IfcFacility,
IfcFacility : [IfcProject, IfcSite, IfcFacility],
IfcRailway : [IfcProject, IfcSite, IfcRailway]}
```

Imagine the entity in question is IfcRailway which is clearly a IfcRailway but also a IfcFacility. 
The dicts in Python are not ordered in some Python versions + the .csv list in not ordered on input either.

Reply:
> in favor of having a more deterministic behavior of the rule

+1 from me. 

---

> aggregated to other IfcSpatialStructureElement

expands to 

> aggregated to IfcProject or ...

Note that `IfcProject` **is not** an `IfcSpatialStructureElement`. This is wrong.

---

> on the .csv implemented by the user

Is it planned that a user would be able to change these `.csv`? How do you ensure validity of such input?

---

> Should it be done in inheritance order... ?

I don't have any opinion on this. This seems to me as an implementation specific design choice. In my view, the stricter the rule is, the better the checking results are - see also my first point above.

---

Is this code any different from the one I reviewed in #67 ?

Reply:
> 
> Note that `IfcProject` **is not** an `IfcSpatialStructureElement`. This is wrong.
> 


Fair point. Changed that in the documentation. 



> Is it planned that a user would be able to change these `.csv`? How do you ensure validity of such input?

No. I intended that only the rule developers will be able to change the csv, bad phrasing on my part. 
I meant that if there are some developments to the spatial composition / decomposition tables in the future, it would be quite easy for us (devs) to change the .csv in comparison to changing the python code. 

> I don't have any opinion on this. This seems to me as an implementation specific design choice. In my view, the stricter the rule is, the better the checking results are - see also my first point above.

I'll try to come up with the inheritance order solution.


> 
> Is this code any different from the one I reviewed in #67 ?

Very similar if not the same. I just needed it in both PRs in the same time.

Reply:
Added sorting by inheritance.